bindings/python/iree/runtime/array_interop.py

```diff
@@ -0,0 +1,232 @@
+       and can have the usual things done with it that one expects to be
+       able to do with an ndarray.
+    2. Device resident: The array is just a handle to a device resident
+       Buffer (and BufferView wrapper). Metadata about the array are accessible
+       (shape and dtype) but anything that touches the data cannot be accessed
+       in this state.
+
+  How a device array comes into existence controls how it can transition
+  between these states:
+    * A user can create a DeviceArray explicitly with a device allocator.
+      Such an array will not be implicitly convertible to host accessible,
+      although accessors exist to do so.
+    * When created by the platform with a synchronization policy, then
+      implicit transfer back to the host will trigger appropriate waits and
+      be performed automatically (this is the common case for function return
+      values if not otherwise configured, as an example).
+  """
+
+  def __init__(self,
+               device: HalDevice,
+               buffer_view: HalBufferView,
+               implicit_host_transfer: bool = False,
+               override_dtype=None):
+    self._device = device
+    self._buffer_view = buffer_view
+    self._implicit_host_transfer = implicit_host_transfer
+    self._override_dtype = override_dtype
+
+    # If the array is host accessible, these will be non-None.
+    self._mapped_memory: Optional[MappedMemory] = None
+    self._host_array: Optional[np.ndarray] = None
+
+  def __array__(self, dtype=None):
+    self._transfer_to_host(True)
+    if dtype is None:
+      return self._host_array
+    else:
+      return self._host_array.__array__(dtype)  # pytype: disable=attribute-error
+
+  def __array_function__(self, func, types, args, kwargs):
+    if func in _DEVICE_HANDLED_FUNCTIONS:
+      return _DEVICE_HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+    # Anything else forces a transfer to host and then delegates to the
+    # host array.
+    host_array = self.to_host()
+    return host_array.__array_function__(func, types, args, kwargs)  # pytype: disable=attribute-error
+
+  def __repr__(self):
+    return f"<IREE DeviceArray: shape={np.shape(self)}, dtype={self.dtype}>"
+
+  @property
+  def is_host_accessible(self):
+    """Whether this array is currently host accessible."""
+    return self._host_array is not None
+
+  def to_host(self) -> np.ndarray:
+    self._transfer_to_host(False)
+    return self._host_array
+
+  def _transfer_to_host(self, implicit):
+    if self._host_array is not None:
+      return
+    if implicit and not self._implicit_host_transfer:
+      raise ValueError(
+          "DeviceArray cannot be implicitly transferred to the host: "
+          "if necessary, do an explicit transfer via .to_host()")
+    self._mapped_memory, self._host_array = self._map_to_host()
+
+  def _map_to_host(self) -> Tuple[MappedMemory, np.ndarray]:
+    # TODO: When synchronization is enabled, need to block here.
```

Comment:
(this was all fyi btw just to say I think it'll all fit into what you have here :)

Reply:
Thanks, yeah, my understanding roughly matches yours, and I think how I have this can be elaborated to those cases. It was definitely on my mind to potentially have another layer of "hollowness" to the array when the metadata is unknown or partially unknown. All of that is a bit better to work through with concrete next steps and has a fair bit of nuance to it.